### PR TITLE
Unified connection in notebooks

### PR DESCRIPTION
**Pull request by rajmusuku**
- Load only MSSQL connections for the kernel 'SQL'
- Load all the connections(MSSQL/HADOOP_KNOX) for the big data kernels
- No change for python kernel, just show 'localhost'
- For big data kernels
   - If an MSSQL connection is not a unified connection, show an error message saying 'Connection is not valid for the kernel: {0}. Choose a different one'
   - If an MSSQL connection is a unified connection, update it's corresponding big data end point in sqlops extension.
   - Update bigdata end point to sqlops extension in case of HADOOP_KNOX connection

### CODE REVIEW

**Comment by rajmusuku** [comment]
Finally figured out the serialization of connection profile object. Just pushed changes.

File: src/sql/parts/notebook/models/modelInterfaces.ts [hunk 1]
```diff
@@ -372,6 +372,8 @@ export interface INotebookModel {
 	pushEditOperations(edits: ISingleNotebookEditOperation[]): void;
 
 	getApplicableConnectionProviderIds(kernelName: string): string[];
+
+	validateUnifiedConnection(connection?: IConnectionProfile):Promise<void>;
```
[the remark this answers is not in the document]
Comment (rajmusuku) [reply src/sql/parts/notebook/models/modelInterfaces.ts]: Moved the validation logic into extension code base. Handled error in notebook model.

**Comment by rajmusuku** [comment]
>           Per comment, let's see if we can move to extension and just handle showing error through standard promise rejection or similar path.

@kevcunnane : Pushed the latest changes. Please review it and let me know.

File: src/sql/parts/notebook/models/notebookContexts.ts [hunk 2]
```diff
@@ -54,7 +54,7 @@ export class NotebookContexts {
 	public static async getContextsForKernel(connectionService: IConnectionManagementService, connProviderIds: string[], kernelChangedArgs?: nb.IKernelChangedArgs, profile?: IConnectionProfile): Promise<IDefaultConnection> {
 		let connections: IDefaultConnection = this.DefaultContext;
 		if (!profile) {
-			if (!kernelChangedArgs || !kernelChangedArgs.newValue ||
+		if (!kernelChangedArgs || !kernelChangedArgs.newValue ||
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/notebookContexts.ts]: Seems like the only change here is you accidentally removed a tab that should still be there. Can you add this back in? And the newline at end of file?

File: src/sql/parts/notebook/models/notebookModel.ts [hunk 3]
```diff
@@ -451,7 +451,10 @@ export class NotebookModel extends Disposable implements INotebookModel {
 			let newConnectionProfile = new ConnectionProfile(this.notebookOptions.capabilitiesService, newConnection);
 			this._activeConnection = newConnectionProfile;
 			this.refreshConnections(newConnectionProfile);
-			await this._activeClientSession.updateConnection(this._activeConnection);
+			this._activeClientSession.updateConnection(this._activeConnection.toIConnectionProfile()).catch(() => {
+				let kernelDisplayName = this.getDisplayNameFromSpecName(this.clientSession.kernel.name);
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/notebookModel.ts]: When this error happens, what is the result in terms of user experience? Do they get "select connection" shown?

File: src/sql/platform/connection/common/connectionProfile.ts [hunk 4]
```diff
@@ -158,7 +158,7 @@ export class ConnectionProfile extends ProviderConnectionInfo implements interfa
 			serverName: this.serverName,
 			databaseName: this.databaseName,
 			authenticationType: this.authenticationType,
-			getOptionsKey: undefined,
+			getOptionsKey: this.getOptionsKey,
```
Comment (kevcunnane) [inline src/sql/platform/connection/common/connectionProfile.ts]: Is this used for lookup? Is it intended to be public?

File: src/sql/parts/notebook/models/notebookModel.ts [hunk 5]
```diff
@@ -451,7 +451,10 @@ export class NotebookModel extends Disposable implements INotebookModel {
...
+				this.notifyError(localize('connectionNotValid', 'Connection is not valid for the kernel: {0}. Choose a different one', kernelDisplayName));
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/notebookModel.ts]: As mentioned in the other review, what we should do is capture the error sent back from the extension (which knows what went wrong) and show to the user.

File: src/sql/parts/notebook/models/notebookModel.ts (hunk 3 above)
Comment (kevcunnane) [reply src/sql/parts/notebook/models/notebookModel.ts]: We need to cleanly handle this so that users know it failed and can try again somehow.

Comment (kevcunnane) [reply src/sql/parts/notebook/models/notebookModel.ts]: Yes let's discuss further. I'm concerned about what happens next - will the Attach To dropdown make it clear that a) you're not connected, and b) you should be able to choose a new connection.

File: src/sql/parts/notebook/models/notebookModel.ts [hunk 6]
```diff
@@ -451,7 +451,11 @@ export class NotebookModel extends Disposable implements INotebookModel {
 			let newConnectionProfile = new ConnectionProfile(this.notebookOptions.capabilitiesService, newConnection);
 			this._activeConnection = newConnectionProfile;
 			this.refreshConnections(newConnectionProfile);
-			await this._activeClientSession.updateConnection(this._activeConnection);
+			this._activeClientSession.updateConnection(this._activeConnection.toIConnectionProfile()).catch((error) => {
+				if (error) {
+					this.notifyError(error.message);
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/notebookModel.ts]: Minor: we have a util method to get the error message. I think it's `notebookUtils.getErrorMessage`. This handles cases where it's a string error in addition to Error class

**Pull request merged**